pkg/retriever/bitswaphelpers/preloadcachingstorage.go

```diff
@@ -0,0 +1,379 @@
+	}
+
+	pl, ok := cs.preloads[link]
+	if ok {
+		pl.refCnt--
+		if pl.refCnt <= 0 {
+			delete(cs.preloads, link)
+		}
+	}
+	cs.preloadsLk.Unlock()
+	if !ok {
+		cs.preloadMisses++
+		// load from fetcher, if it can be fetched
+		r, err := cs.fetcher(linkCtx, link)
+		if err != nil {
+			if nf, ok := err.(interface{ NotFound() bool }); ok && nf.NotFound() {
+				cs.preloadsLk.Lock()
+				cs.notFound[string(link.(cidlink.Link).Cid.Hash())] = struct{}{}
+				cs.preloadsLk.Unlock()
+			}
+			return nil, err
+		}
+		// write to parent and cache and return
+		return loadTo(
+			r,
+			[]linking.BlockWriteOpener{cs.cacheLinkSystem.StorageWriteOpener, cs.parentLinkSystem.StorageWriteOpener},
+			linkCtx,
+			link)
+	}
+	cs.preloadingHits++
+	cs.preloadLink(pl, linkCtx, link)
+	select {
+	case <-linkCtx.Ctx.Done():
+		return nil, linkCtx.Ctx.Err()
+	case <-pl.loaded:
+		if pl.err != nil {
+			return nil, pl.err
+		}
+		// TODO: delete from storage if/when possible?  if pl.refCnt <= 0 {}
+		// load from cache, write direct to parent and return
+		r, err := linkSystemGetStream(cs.cacheLinkSystem, linkCtx, link)
+		if err != nil {
+			return nil, err
+		}
+		return loadTo(
+			r,
+			[]linking.BlockWriteOpener{cs.parentLinkSystem.StorageWriteOpener},
+			linkCtx,
+			link)
+	}
+}
+
+func loadTo(
+	reader io.Reader,
+	destination []linking.BlockWriteOpener,
+	linkCtx linking.LinkContext,
+	link ipld.Link) (io.Reader, error) {
+
+	blc := make([]linking.BlockWriteCommitter, 0, len(destination))
+	for _, blo := range destination {
+		writer, c, err := blo(linkCtx)
+		if err != nil {
+			return nil, err
+		}
+		blc = append(blc, c)
+		reader = io.TeeReader(reader, writer)
+	}
+	byts, err := io.ReadAll(reader) // slurp it in so the writers can commit
+	if err != nil {
+		return nil, err
+	}
+	err = nil
+	for _, c := range blc {
+		err = multierr.Append(err, c(link))
+	}
+	return bytes.NewBuffer(byts), err
+}
+
+func (cs *PreloadCachingStorage) preloadLink(pl *preloadingLink, linkCtx linking.LinkContext, link ipld.Link) {
+	pl.loadSyncer.Do(func() {
+		defer close(pl.loaded)
+		reader, err := cs.fetcher(linkCtx, link)
+		if err != nil {
+			if nf, ok := err.(interface{ NotFound() bool }); ok && nf.NotFound() {
+				cs.preloadsLk.Lock()
+				cs.notFound[string(link.(cidlink.Link).Cid.Hash())] = struct{}{}
+				cs.preloadsLk.Unlock()
+			}
+			pl.err = err
+		} else {
+			w, c, err := cs.cacheLinkSystem.StorageWriteOpener(linkCtx)
+			if err != nil {
+				pl.err = err
+				return
+			}
+			if _, err := io.Copy(w, reader); err != nil {
+				pl.err = err
+				return
+			}
+			if err := c(link); err != nil {
+				pl.err = err
+				return
+			}
+		}
+	})
+}
+
+func (cs *PreloadCachingStorage) Start(ctx context.Context) {
+	if cs.cancel != nil {
+		panic("already started")
+	}
+	ctx, cs.cancel = context.WithCancel(ctx)
+	go cs.run(ctx)
+}
+
+func (cs *PreloadCachingStorage) run(ctx context.Context) {
+	feed := make(chan *request)
+	defer func() {
+		//close(feed)
+	}()
```

Comment:
unused?

Reply:
oh, actually I think that should be used! thanks